Data_Engine: Adding create methods and first query methods for Table class

IsakNaslundBh opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->

### NOTE: Depends on 
<!-- Link to any additional PRs in other repos required for this PR to function -->
<!-- Delete if not required -->

https://github.com/BHoM/BHoM/pull/595
   
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #1285 

<!-- Add short description of what has been fixed -->

Adding create methods and first Query methods for Tables.

Please note that the DataTable can not be serialised to Bson by default. I have code ready for this, but will need to be aligned with the work done by @adecler in https://github.com/BHoM/BHoM_Engine/pull/1270


### Test files
<!-- Link to test files to validate the proposed changes -->

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM_Engine/Data_Engine/Issue_1289_CreateTables?csf=1&e=KcLoUo

### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->

- Adding Create method for table class. One from `List<IBHoMObject>` and one from 2D table with header values.
- Query method for getting table rows matching exact values in table
- Query method for getting table rows matching values above set axis-values

### Additional comments
<!-- As required -->

@Martian42

epignatelli commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,139 @@
+
+        /***************************************************/
+
+        [Description("")]
```
Are these going to be filled? If not, we can remove them?

IsakNaslundBh replied:
Thanks @epignatelli , yes, need to get all of these filled in.

epignatelli commented on Data_Engine/Query/Axes.cs:
```diff
@@ -0,0 +1,56 @@
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("")]
```
As above

epignatelli commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,139 @@
+        [Description("")]
+        [Input("", "")]
+        [Output("", "")]
+        public static Table Table<T>(string name, string axis1Name, List<IComparable> axis1Values, string axis2Name, List<IComparable> axis2Values, List<List<T>> values, string valuesName = "Values")
```
What is an `IComparable` exactly? Can we input it from the UI side of things?
Not commenting all the rest, it's in other places as well.

IsakNaslundBh commented:
@adecler , would not mind a review from you on this, especially on the Serialiser_Engine part.

adecler commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,152 @@
+            foreach (var kvp in properties)
+            {
+                
+                columns.Add(new DataColumn(kvp.Key, kvp.Value.GetType()));
```
This needs a null check:

![image](https://user-images.githubusercontent.com/16853390/68446078-a04a9580-0216-11ea-9538-deb0d3ae03c6.png)

IsakNaslundBh commented:
Pinging this!

Would anyone please be able to have a look at this PR so we can get it merged. Thanks!

FraserGreenroyd commented:
Sorry @IsakNaslundBh yes I'll try and get on this today!

FraserGreenroyd commented:
/azp run BHoM.CheckCore

FraserGreenroyd commented:
/azp run BHoM_Engine.CheckCore

FraserGreenroyd commented:
> No pipelines are associated with this pull request.

Good to know, thanks bot 🤦‍♂

FraserGreenroyd commented on Data_Engine/Query/Axes.cs:
```diff
@@ -0,0 +1,56 @@
+        [Output("axes", "The table axes names")]
+        public static List<string> Axes(this Table table)
+        {
+            var columnNames = new List<string>();
```
Could we not simplify this with a Linq query instead?
`return table.Data.Columns.Select(x => x.ColumnName).ToList();`

IsakNaslundBh replied:
Tried that at first, but can not use Linq on table.Data.Columns for whatever reason. Not sure why atm.

Could do:  `            return table.Data.Columns.Cast<DataColumn>().Select(x => x.ColumnName).ToList();` if you think that is cleaner?

FraserGreenroyd commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,194 @@
+                new DataColumn(axis2Name, axis2Values.First().GetType()),
+                new DataColumn(valuesName, typeof(T))
+            };
+
```
Excess whitespace

FraserGreenroyd commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,194 @@
+
+
+            table.Columns.AddRange(cols);
+
```
Excess whitespace

FraserGreenroyd commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,194 @@
+            }
+
+            return new Table() { Data = table, Name = name };
+            
```
Excess whitespace

FraserGreenroyd commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,194 @@
+
+            return new Table() { Data = table, Name = name };
+            
+
```
Excess whitespace

FraserGreenroyd commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,194 @@
+            }
+            return dic;
+        }
+    
```
Excess whitespace

al-fisher commented:
> No pipelines are associated with this pull request.

>Good to know, thanks bot 🤦‍♂


@FraserGreenroyd it is the beginning of the end! 🤖 
😆

FraserGreenroyd commented on Data_Engine/Create/Table.cs:
```diff
@@ -0,0 +1,194 @@
+                dic["Tags"] = obj.Tags.GetType();
+
+            return dic;
+
```
Excess whitespace

FraserGreenroyd commented:
> > No pipelines are associated with this pull request.
> 
> > Good to know, thanks bot 🤦‍♂
> 
> @FraserGreenroyd it is the beginning of the end! 🤖
> 😆

Skynet here we come... time to build a bunker! *tinfoil hat firmly placed*

FraserGreenroyd commented:
Test script worked with some modifications (I think the GH file was slightly out of date to the latest code, but replacing the components worked fine for me), so am happy to approve this PR when the above code comments are addressed @IsakNaslundBh :smile:

IsakNaslundBh commented:
Thanks @FraserGreenroyd 🚀 

Have fixed your white space comments. Also added a method to get all data out from the table. Added that method to the testscript as well and updated the outdated components.

FraserGreenroyd commented:
/azp run BHoM_Engine.CheckInstaller
